web/app/services/screensaver.service.js

```diff
@@ -0,0 +1,349 @@
+                saveSettings();
+                $interval.cancel(_slideshowTimer);
+                _slideshowTimer = null;
+                log("Stopping slideshow. No dashboard found.");
+                $location.url("/view/ " + _config.onStart.dashboard);
+                return;
+            }
+
+            var nextDbId = _slideshowDashboards[_currentDbIndex].id;
+            if (!dashboardExists(nextDbId)) {
+                removeDashboard(nextDbId);
+                nextDashboard();
+                return;
+            }
+            _currentDbIndex = ++_currentDbIndex < _slideshowDashboards.length ? _currentDbIndex : 0;
+            $location.url("/view/" + nextDbId);
+        }
+
+        var slideshow = function () {
+            log("Screensaver (" + _config.onStart.type + ") started in dashboard " + $route.current.params.i);
+            _currentDbIndex = 0;
+            nextDashboard();
+            _slideshowTimer = $interval(nextDashboard, (_config.slideshowIntervalSec || 10) * 1000);
+        }
+
+        var start = function () {
+            if (_isRunning)
+                return;
+
+            _isRunning = true;
+            watchEvents();
+
+            if (_config.onStart.type === 'slideshow') {
+                slideshow();
+            } else {
+                $location.url("/view/" + _config.onStart.dashboard)
+            }
+        };
+
+        var stop = function (isFromOHService) {
+            if (!_isRunning)
+                return;
+
+            unWatchEvents();
+            $interval.cancel(_slideshowTimer);
+            _slideshowTimer = null;
+            _isRunning = false;
+            log("Screensaver stopped.");
+
+            if (_config.isEnabled) {
+                idleTimerStart();
+            }
+
+            if (isFromOHService)
+                return;
+
+            if (_config.onStop.type === 'gotodashboard') {
+                $location.url("/view/" + _config.onStop.dashboard);
+            }
+        };
+
+        var idleTimerStart = function () {
+            _idleTimer = $timeout(onIdle, _config.idleTimeoutSec * 1000);
+        }
+
+        var idleTimerStop = function () {
+            $timeout.cancel(_idleTimer);
+            _idleTimer = null;
+        }
+
+        var init = function () {
+            $timeout(function () {
+                initConfig();
+                if (!_config || !_config.idleTimeoutSec || !_config.isEnabled)
+                    return;
+                idleTimerStart();
+            });
+        };
+
+        var saveSettings = function (config) {
+            config = config || _config;
+            // Uniqify our arrays
+            function onlyUnique(value, index, self) { return self.indexOf(value) === index; }
+            config.onStart.dashboards = config.onStart.dashboards.filter(onlyUnique);
+            config.onStart.dashboardsExcluded = config.onStart.dashboardsExcluded.filter(onlyUnique);
```

Comment:
Why store the *excluded* dashboards in the config, wouldn't the ones in the rotation be enough? Seems weird and would need maintenance when the dashboard list is updated.
This is what I have in the panel configuration after storing the settings?
```
    "settings": {
        ...
        "screensaver": {
            "idleTimeoutSec": 300,
            "slideshowIntervalSec": 5,
            "isEnabled": true,
            "eventsToWatch": {
                "mousedown": true,
                "keydown": true,
                "mousewheel": true,
                "touchstart": true,
                "touchmove": true
            },
            "additionalEventsToWatch": [],
            "onStart": {
                "type": "slideshow",
                "dashboardsExcluded": [
                    {
                        "id": "test"
                    },
                    {
                        "id": "mobile"
                    },
                    {
                        "id": "map"
                    },
                    {
                        "id": "weather"
                    },
                    {
                        "id": "ec  vdsgsrg"
                    }
                ],
                "dashboards": [],
                "dashboard": null
            },
            "onStop": {
                "type": "stop",
                "dashboard": "test"
            }
        }
    }
```

Reply:
I initially wanted to have 2 containers for the dbs (included/excluded) so that I would not have to loop through the array while in the settings controller (for the included/excluded) draggable items. But let me see if I can improve this.